src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -161,9 +161,38 @@ public HostOutputMessage HostOutputMessage
                 await _functionInstanceLogger.DeleteLogFunctionStartedAsync(functionStartedMessageId, cancellationToken);
             }
 
+            await ThrowIfTimeoutOrFatal(exceptionInfo, _timeoutConfig.TimeoutGracePeriod, _backgroundExceptionDispatcher);
+
             return exceptionInfo != null ? new ExceptionDispatchInfoDelayedException(exceptionInfo) : null;
         }
 
+        internal static async Task ThrowIfTimeoutOrFatal(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod, IWebJobsExceptionHandler exceptionHandler)
```

Comment:
nit: I'd rename this HandleExceptionAsync, and only call it above if exceptionInfo != null